### ISSUE TITLE
Vanilla type error leads to "Unknown assertion failed"

### ISSUE THREAD

**Issue by owenarden** [FStarLang/FStar#446]
Hi,

The following code fails to compile the "fails" function saying "Unknown assertion failed", but the "works" function seems ok.  Is this a bug?

```
module Test

kind Pre (context:Type) = STPre_h context
kind Post (context:Type) (a:Type) = STPost_h context a
kind WP (context:Type) (a:Type) = STWP_h context a

effect P (a:Type) (wp:PureWP a) = PURE a wp wp
type M (context:Type) (a:Type) (wp:WP context a) = 
  ctx0:context -> P (a * context) (fun 'q -> wp (fun a context -> 'q (a, context)) ctx0)

type ctx = int

let works (#a:Type) (#wp1:WP ctx a) (f:M ctx a wp1) : bool = true

let fails (#a:Type) (#b:Type) (#wp1:WP ctx a) (#wp2:WP ctx b)
  (f:M ctx a wp1) (g:M ctx a wp2) : bool = true
```

**Comment by catalin-hritcu** [comment]
The error message is indeed horrible, so it's good that you reported this, we should try to improve that. The main problem is in the code of `fails` though: the type of `g` is wrong, it should be `M ctx b wp2` to match the type of `wp2`.

**Comment by owenarden** [comment]
I just noticed that I had substituted a for b also, sorry about that. Thanks for taking a look.

I seemed to get slightly better error messages when (1) I wrote an explicit type for "fails" (2) copied the all.fst module into the local directory and opened FStar.All and (3) in response to an error message, changed the return type of "fails" to GTot bool.  

I understand why (1) was useful, but (2) and (3) were kind of blindly groping in the dark.  I suspect that they are related to the fact I was trying to use (the actual version of) "fails" in a refinement type, but it's weird that I had to get all the way through that chain of errors before I was forced to notice my obvious a vs b error.
